talk/expert/cpp20concepts.tex

```diff
@@ -0,0 +1,194 @@
+    \end{exampleblock}
+  \end{frame}
+
+  \begin{frame}[fragile]
+    \frametitlecpp[20]{C++20 Requirements and Concepts}
+    \begin{block}{Concepts}
+      When a given set of requirements may be reused often, one should gather them in a concept.
+    \end{block}
+    \begin{exampleblock}{With {\it concept}}
+      \begin{cppcode*}{}
+      template< typename T>
+      concept MyFloatingPoint =
+        (std::is_floating_point_v<T>) &&
+        (std::numeric_limits<T>::epsilon()>0) ;
+
+      template<typename T>
+      requires MyFloatingPoint<T>
+      bool equal( T e1, T e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{exampleblock}
+  \end{frame}
+
+  \begin{frame}[fragile]
+    \frametitlecpp[20]{C++20 Requirements and Concepts}
+    \begin{block}{Within the template head}
+        Actually, when relevant, the concept can be directly used instead of {it typename} within the template head.
+    \end{block}
+    \begin{exampleblock}{Instead of {\it typename}}
+      \begin{cppcode*}{}
+      template< typename T>
+      concept MyFloatingPoint =
+        (std::is_floating_point_v<T>) &&
+        (std::numeric_limits<T>::epsilon()>0) ;
+
+      template<MyFloatingPoint T>
+      bool equal( T e1, T e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{exampleblock}
+  \end{frame}
+
+  \begin{frame}[fragile]
+    \frametitlecpp[20]{C++20 Requirements and Concepts}
+    \begin{block}{With abbreviated function templates}
+        The concepts can be used together with {\it auto} in the abbreviated function templates.
+    \end{block}
+    \begin{exampleblock}{With {\it auto}}
+      \begin{cppcode*}{}
+      template< typename T>
+      concept MyFloatingPoint =
+        (std::is_floating_point_v<T>) &&
+        (std::numeric_limits<T>::epsilon()>0) ;
+
+      bool equal( MyFloatingPoint auto e1, MyFloatingPoint auto e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{exampleblock}
+  \end{frame}
+
+  \begin{frame}[fragile]
+    \frametitlecpp[20]{C++20 Requirements and Concepts}
+    \begin{block}{Standard concepts}
+        Writing a bug-proof concept is actually is really expert task.
+        Whenever you can, use the ones provided by the standard library.
+        Not surprinsgly, there is one for floating point numbers.
+    \end{block}
+    \begin{exampleblock}{With {\it std::}}
+      \begin{cppcode*}{}
+      #include <concepts>
+
+      bool equal( std::floating_point auto e1, std::floating_point auto e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{exampleblock}
+  \end{frame}
+
+  \begin{frame}[fragile]
+    \frametitlecpp[20]{C++20 Requirements and Concepts}
+    \begin{block}{Within a function body}
+    Concepts are usable wherever a boolean is expected, including {\it if constexpr}, because they are evaluated at compile-time.
```

Comment:
```suggestion
    Concepts are also usable wherever a boolean is expected, including {\it if constexpr}, because they are evaluated at compile-time.
```